tree-sitter-stack-graphs/src/lib.rs

```diff
@@ -0,0 +1,550 @@
+//!     type = "push",
+//!     symbol = (source-text @id),
+//!     scope = new_exported_scope_node
+//! }
+//! ```
+//!
+//! To make a _pop scoped symbol_ node, you must provide a `scoped` attribute, whose value must be
+//! `#true`.  (You don't know in advance which particular scope stack will be popped off.) For
+//! instance:
+//!
+//! ``` skip
+//! (identifier) @id {
+//!   node new_pop_scoped_symbol_node
+//!   attr (new_pop_scoped_symbol_node)
+//!     type = "pop",
+//!     symbol = (source-text @id),
+//!     scoped = #true
+//! }
+//! ```
+//!
+//! ### Annotating nodes with location information
+//!
+//! You can annotate any stack graph node that you create with location information, identifying
+//! the portion of the source file that the node “belongs to”.  This is _required_ for `definition`
+//! and `reference` nodes, since the location information determines which parts of the source file
+//! the user can “click on”, and the “destination” of any code navigation queries the user makes.
+//! To do this, add a `source_node` attribute, whose value is a syntax node capture:
+//!
+//! ``` skip
+//! (function_definition name: (identifier) @id) @func {
+//!   node def
+//!   attr (def) type = "definition", symbol = (source-text @id), source_node = @func
+//! }
+//! ```
+//!
+//! Note how in this example, we use a different syntax node for the “target” of the definition
+//! (the entirety of the function definition) and for the _name_ of the definition (the content of
+//! the function's `name`).
+//!
+//! ### Connecting stack graph nodes with edges
+//!
+//! To connect two stack graph nodes, use the `edge` statement to add an edge between them:
+//!
+//! ``` skip
+//! (function_definition name: (identifier) @id) @func {
+//!   node def
+//!   attr (def) type = "definition", symbol = (source-text @id), source_node = @func
+//!   node body
+//!   edge def -> body
+//! }
+//! ```
+//!
+//! To implement shadowing (where later definitions of the same name in the same scope “replace” or
+//! “shadow” earlier ones), you can add a `precedence` attribute to each edge to indicate which
+//! paths are prioritized:
+//!
+//! ``` skip
+//! (function_definition name: (identifier) @id) @func {
+//!   node def
+//!   attr (def) type = "definition", symbol = (source-text @id), source_node = @func
+//!   node body
+//!   edge def -> body
+//!   attr (def -> body) precedence = 1
+//! }
+//! ```
+//!
+//! (If you don't specify a `precedence`, the default is 0.)
+//!
+//! ### Referring to the singleton nodes
+//!
+//! The _root node_ and _jump to scope node_ are singleton nodes that always exist for all stack
+//! graphs.  You can refer to them using the `ROOT_NODE` and `JUMP_TO_SCOPE_NODE` global variables:
+//!
+//! ``` skip
+//! (function_definition name: (identifier) @id) @func {
+//!   node def
+//!   attr (def) type = "definition", symbol = (source-text @id), source_node = @func
+//!   edge (ROOT_NODE -> def)
+//! }
+//! ```
```

Comment:
I see you have adopted different syntax than what `tree-sitter-code-nav` uses. I am wondering about the ergonomics of it, as it seems quite a bit more verbose.

For example, `type = "definition", symbol = (source-text @id)` used to be simply `definition = @id`. (I suppose this example even implied `source_node = @id`.) I definitely support the possibility to differentiate all of these, but I am wondering if we can retain lean syntax for the most common cases.

We can stay closer to the old syntax by dropping `type` for `definition`, `drop`, etc attributes. This looks more semantic (I expect quoted strings to allow arbitrary content), and is more compact. If we feel like `symbol` is really the main (or even only) value associated with those node, I think doing `definition =` makes sense as well. For nodes without symbols, the implied `#true` should to fine.

The final question is, whether we we would like to allow `definition = @capture` and have `(source-text @capture)` and `source_node = @capture` implied from that? I'm leaning towards yes, since this is such a common pattern.

Reply:
My thinking here is twofold.

First, I was trying to make the attribute vocabulary more robust.  If we have attributes like `definition = [...]` and `push_scope = [...]`, then it's possible for the graph DSL author to add more than one of those attributes to a node.  We can definitely add checks for that in this Rust crate, but with `type = {definition|push|...}`, it's not possible to construct that particular erroneous pattern.

Second, I hear you that the ergonomics might be worse — but I was thinking that once we add "macros" or "procedures" or whatever we want to call them to the graph DSL, _that's_ how we would make this more ergonomic.  If there's a procedure named `add_definition_node` that you call, that decouples the way that we make it easy to use in the graph DSL from the particular attributes and values that we use under the covers.